File: pi-agent/src/split-registry.mjs
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

export const SPLIT_REGISTRY_PATH = fileURLToPath(
  new URL("../../run_on_benchmark/splits/thesis_split_v2.json", import.meta.url),
);

let cached = null;

export function loadSplitRegistry(path = SPLIT_REGISTRY_PATH) {
  if (path === SPLIT_REGISTRY_PATH && cached) return cached;
  const value = JSON.parse(readFileSync(path, "utf8"));
  if (value.schema_version !== 2) throw new Error(`unsupported split registry: ${path}`);
  if (path === SPLIT_REGISTRY_PATH) cached = value;
  return value;
}

export function splitRegistrySha256(path = SPLIT_REGISTRY_PATH) {
  return createHash("sha256").update(readFileSync(path)).digest("hex");
}

function matches(caseId, member) {
  if (caseId === member) return true;
  const range = /^([a-zA-Z_-]+):(\d+)-(\d+)$/.exec(member);
  const item = /^([a-zA-Z_-]+)-(\d+)$/.exec(caseId);
  if (!range || !item) return false;
  return range[1].toLowerCase() === item[1].toLowerCase()
    && Number(item[2]) >= Number(range[2])
    && Number(item[2]) <= Number(range[3]);
}

export function assignedSplit(benchmarkId, caseId, registry = loadSplitRegistry()) {
  const splits = registry.benchmarks?.[benchmarkId]?.splits ?? {};
  const found = Object.entries(splits)
    .filter(([, members]) => members.some((member) => matches(caseId, member)))
    .map(([split]) => split);
  if (found.length > 1) {
    throw new Error(`case belongs to multiple splits: ${benchmarkId}/${caseId}: ${found.join(", ")}`);
  }
  return found[0] ?? null;
}

export function assertCaseSplit(benchmarkId, caseId, requestedSplit, registry) {
  const actual = assignedSplit(benchmarkId, caseId, registry);
  if (!actual) throw new Error(`case is not assigned in frozen split registry: ${benchmarkId}/${caseId}`);
  if (actual !== requestedSplit) {
    throw new Error(
      `split mismatch for ${benchmarkId}/${caseId}: requested ${requestedSplit}, frozen as ${actual}`,
    );
  }
  return actual;
}

File: run_on_benchmark/split_registry.py
"""Frozen case membership checks for thesis experiments."""

from __future__ import annotations

import hashlib
import json
import re
from functools import lru_cache
from pathlib import Path
from typing import Any

REGISTRY_PATH = Path(__file__).with_name("splits") / "thesis_split_v2.json"


@lru_cache(maxsize=1)
def load_split_registry() -> dict[str, Any]:
    value = json.loads(REGISTRY_PATH.read_text(encoding="utf-8"))
    if value.get("schema_version") != 2:
        raise ValueError(f"unsupported split registry: {REGISTRY_PATH}")
    return value


def split_registry_sha256() -> str:
    return hashlib.sha256(REGISTRY_PATH.read_bytes()).hexdigest()


def _matches(case_id: str, member: str) -> bool:
    if member == case_id:
        return True
    match = re.fullmatch(r"([a-zA-Z_-]+):(\d+)-(\d+)", member)
    case_match = re.fullmatch(r"([a-zA-Z_-]+)-(\d+)", case_id)
    if not match or not case_match:
        return False
    prefix, start, end = match.groups()
    case_prefix, number = case_match.groups()
    return prefix.casefold() == case_prefix.casefold() and int(start) <= int(number) <= int(end)


def assigned_split(benchmark_id: str, case_id: str) -> str | None:
    benchmark = (load_split_registry().get("benchmarks") or {}).get(benchmark_id) or {}
    matches = [
        split for split, members in (benchmark.get("splits") or {}).items()
        if any(_matches(case_id, member) for member in members)
    ]
    if len(matches) > 1:
        raise ValueError(f"case belongs to multiple splits: {benchmark_id}/{case_id}: {matches}")
    return matches[0] if matches else None


def assert_case_split(benchmark_id: str, case_id: str, requested_split: str) -> str:
    actual = assigned_split(benchmark_id, case_id)
    if actual is None:
        raise ValueError(f"case is not assigned in frozen split registry: {benchmark_id}/{case_id}")
    if actual != requested_split:
        raise ValueError(
            f"split mismatch for {benchmark_id}/{case_id}: requested {requested_split}, frozen as {actual}"
        )
    return actual
